samza-core/src/main/java/org/apache/samza/classloader/IsolatingClassLoaderFactory.java

```diff
@@ -0,0 +1,195 @@
+    String baseDirectoryPath = System.getProperty("user.dir");
+    File apiLibDirectory =
+        libDirectory(new File(baseDirectoryPath, IsolationUtils.APPLICATION_MASTER_API_DIRECTORY));
+    LOG.info("Using API lib directory: {}", apiLibDirectory);
+    File infrastructureLibDirectory =
+        libDirectory(new File(baseDirectoryPath, IsolationUtils.APPLICATION_MASTER_INFRASTRUCTURE_DIRECTORY));
+    LOG.info("Using infrastructure lib directory: {}", infrastructureLibDirectory);
+    File applicationLibDirectory =
+        libDirectory(new File(baseDirectoryPath, IsolationUtils.APPLICATION_MASTER_APPLICATION_DIRECTORY));
+    LOG.info("Using application lib directory: {}", applicationLibDirectory);
+
+    ClassLoader apiClassLoader = buildApiClassLoader(apiLibDirectory);
+    ClassLoader applicationClassLoader =
+        buildApplicationClassLoader(applicationLibDirectory, apiLibDirectory, apiClassLoader);
+
+    // the classloader to ultimately return is the one with the infrastructure classpath
+    return buildInfrastructureClassLoader(infrastructureLibDirectory, apiLibDirectory, apiClassLoader,
+        applicationClassLoader);
+  }
+
+  private static ClassLoader buildApiClassLoader(File apiLibDirectory) {
+    // null parent means to use bootstrap classloader as the parent
+    return new URLClassLoader(getClasspathAsURLs(apiLibDirectory), null);
+  }
+
+  private static ClassLoader buildApplicationClassLoader(File applicationLibDirectory, File apiLibDirectory,
+      ClassLoader apiClassLoader) {
+    return LoaderBuilder.anIsolatingLoader()
+        // look in application directory for JARs
+        .withClasspath(getClasspathAsURIs(applicationLibDirectory))
+        // getClasspathAsURIs should already satisfy this, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(applicationLibDirectory, false))
+        .withParentRelationship(buildApiParentRelationship(apiLibDirectory, apiClassLoader))
+        .build();
+  }
+
+  private static ClassLoader buildInfrastructureClassLoader(File infrastructureLibDirectory, File apiLibDirectory,
+      ClassLoader apiClassLoader, ClassLoader applicationClassLoader) {
+    return LoaderBuilder.anIsolatingLoader()
+        // look in application directory for JARs
+        .withClasspath(getClasspathAsURIs(infrastructureLibDirectory))
+        // getClasspathAsURIs should already satisfy this, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(infrastructureLibDirectory, false))
+        .withParentRelationship(buildApiParentRelationship(apiLibDirectory, apiClassLoader))
+        // need to be able to fall back to application for loading certain classes
+        .addFallbackDelegate(buildFallbackDelegateRelationship(applicationClassLoader))
+        .build();
+  }
+
+  private static DelegateRelationship buildApiParentRelationship(File apiLibDirectory, ClassLoader apiClassLoader) {
```

Comment:
Can you add javadoc explaining what this method does. E.g., `Sets up classloader relationships for the API classloader such that ... bootstrap classes (e.g. java.lang Classes) are loaded from ... and api classes (as specified by classes listed in file ....) are loaded from ...` or something to that effect.

Reply:
Same for the buildApplicationClassLoader method and buildInfrastructureClassLoader method. Would be good to explain what a `parentRelationship` entails, e.g., who is parent of whom, and what are the implications for preference when loading classes. The API naming doesn't make it clear enough.

Reply:
Would be even better to include an ascii diagram explaining the relationship (or a link to one in the SEP).